pkg/mutation/schema/schema.go

```diff
@@ -0,0 +1,387 @@
+		panic(err)
+	}
+}
+
+// Remove removes the mutator with the given id from the
+// db.
+func (db *DB) Remove(id types.ID) {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+	db.remove(id)
+	delete(db.mutators, id)
+}
+
+func (db *DB) remove(id types.ID) {
+	mutator, ok := db.mutators[id]
+	if !ok {
+		// no mutator found, nothing to do
+		return
+	}
+	bindings := mutator.SchemaBindings()
+	gvks := map[schema.GroupVersionKind]struct{}{}
+	// deduplicate GVKs
+	for _, binding := range bindings {
+		for _, group := range binding.Groups {
+			for _, version := range binding.Versions {
+				for _, kind := range binding.Kinds {
+					gvk := schema.GroupVersionKind{
+						Group:   group,
+						Version: version,
+						Kind:    kind,
+					}
+					gvks[gvk] = struct{}{}
+				}
+			}
+		}
+	}
+
+	for gvk := range gvks {
+		s, ok := db.schemas[gvk]
+		if !ok {
+			log.Error(nil, "mutator associated with missing schema", "mutator", id, "schema", gvk)
+			panic(fmt.Sprintf("mutator %v associated with missing schema %v", id, gvk))
+		}
+		s.remove(mutator.Path().Nodes)
+		if s.root == nil {
+			delete(db.schemas, gvk)
+		}
+	}
+}
+
+type scheme struct {
+	gvk  schema.GroupVersionKind
+	root *node
+}
+
+func (s *scheme) add(ref []parser.Node) error {
+	if s.root == nil {
+		s.root = &node{}
+	}
+	return s.root.add(ref)
+}
+
+func (s *scheme) remove(ref []parser.Node) {
+	if s.root == nil {
+		return
+	}
+	s.root.remove(ref)
+	if s.root.referenceCount == 0 {
+		s.root = nil
+	}
+}
+
+type node struct {
+	referenceCount uint
+	nodeType       parser.NodeType
+
+	// list-type nodes have a key field and only one child
+	keyField *string
+	child    *node
+
+	// object-type nodes only have children
+	children map[string]*node
+}
+
+// backup creates a shallow copy of the node we can restore in case of error
+func (n *node) backup() *node {
+	return &node{
+		referenceCount: n.referenceCount,
+		nodeType:       n.nodeType,
+		keyField:       n.keyField,
+	}
+}
+
+func (n *node) restore(backup *node) {
+	// we assume reference count is always incremented, so we can avoid storing it
+	// in the stack
+	n.referenceCount = backup.referenceCount
+	n.nodeType = backup.nodeType
+	n.keyField = backup.keyField
+}
+
+func (n *node) add(ref []parser.Node) error {
+	backup := n.backup()
+	n.referenceCount++
+
+	// we should stop recursing when len(ref) == 1, but
+	// this guards against infinite recursion in the case
+	// where there is a bug in the switch code below
+	if len(ref) == 0 {
+		return nil
+	}
+
+	current := ref[0]
+	switch t := current.Type(); t {
+	case parser.ObjectNode:
+		if n.nodeType != "" && n.nodeType != parser.ObjectNode {
+			return fmt.Errorf("node type conflict: %v vs %v", n.nodeType, parser.ObjectNode)
+		}
+		n.nodeType = parser.ObjectNode
+		obj := current.(*parser.Object)
+		if n.children == nil {
+			n.children = make(map[string]*node)
+		}
+		if len(ref) == 1 {
```

Comment:
Why not just checking for len == 1 on the top (or-ed with 0) ?

Reply:
I do it here so we can process list nodes, setting expectations like `keyField`.

I could populate these when I make the child node, but then I'd need to introspect the type information for the child.

Deferring the populating of these expectations, I avoid some extra type checking logic at the expense of moving the base-case-check to a less-normal part of the code.

Reply:
Got it, thanks